message-signing/java/signature-verification/src/main/java/com/starlingbank/StarlingMessageSignatureValidator.java

```diff
@@ -0,0 +1,240 @@
+
+    // Get the Authorization header and split into required elements.
+    // Note that for the purposes of this tool we are simplifying this process as we do not require the
+    // Bearer token or the keyid, though both of these elements are necessary for requests to the Starling API.
+    Map<String, String> headerMap = null;
+    try {
+      String authHeader = request.headers("Authorization");
+      headerMap = Splitter.on(',')
+          .withKeyValueSeparator(Splitter.on('=').limit(2))
+          .split(authHeader);
+    } catch (Exception e) {
+      System.err
+          .println("ERROR: Could not process Authorization header, aborting validation check.");
+      halt(400, "ERROR: Could not process Authorization header, aborting validation check.");
+    }
+
+    // Parse the Authorization Header to get required sub elements.
+    String algorithm = getHeader("algorithm", headerMap);
+    String signature = getHeader("signature", headerMap);
+    String headers = getHeader("headers", headerMap);
+
+    // Check that the "headers" element from the Authorization header contains all the required elements (Date, Digest and (request-target).
+    List<String> headersList = Splitter.on(" ").splitToList(headers.toLowerCase());
+    if (!headersList.containsAll(REQUIRED_HEADERS)) {
+      SortedSet<String> missingHeaders = Sets.newTreeSet(REQUIRED_HEADERS);
+      missingHeaders.removeAll(headersList);
+      responseBody.add(
+          "Validation failure: authHeader is missing some required headers: " + Joiner.on(", ")
+              .join(missingHeaders) + "\n");
+    }
+
+    // Get the values for each of the signed headers listed in the Authorization header and build into a single string for verification of the Signature.
+    List<String> lines = Splitter.on(" ").splitToList(headers).stream()
+        .map(header -> {
+          String value = header.equals("(request-target)") ? requestPath : request.headers(header);
+          return header + ": " + value;
+        })
+        .collect(Collectors.toList());
+    String signingString = Joiner.on("\n").join(lines);
+
+    // For non-GET requests, verify that the Message Digest supplied in the request headers matches the body payload.
+    if (!request.requestMethod().equalsIgnoreCase("GET")) {
+      String payloadDigest = Base64.getEncoder()
+          .encodeToString(MessageDigest.getInstance("SHA-512").digest(request.bodyAsBytes()));
+      if (request.headers("Digest").equals(payloadDigest)) {
+        responseBody.add("Successfully verified digest against payload.\n");
+      } else {
+        responseBody.add("Digest could not be verified against payload.\n");
+      }
+    }
+
+    responseBody.add("Looking for publicKey file in:");
+    responseBody.add(new File(publicKeyPath).getAbsolutePath() + "\n");
+    System.out.println("Looking for publicKey file in:");
+    System.out.println(new File(publicKeyPath).getAbsolutePath() + "\n");
```

Comment:
Let's front-load the key parsing in `main`. The tool isn't useful without it and it's better to just create the `Key` once.